src/test/scala/io/iohk/ethereum/network/TimeSlotStatsSpec.scala

```diff
@@ -0,0 +1,286 @@
+      stats.get("foo") shouldBe 3
+      stats.get("bar") shouldBe 6
+      stats.get("BAR") shouldBe 0
+    }
+  }
+
+  it should "aggregate all stats" in testAggregate {
+    for {
+      stats <- getStats
+    } yield {
+      stats.getAll() shouldBe Map("foo" -> 3, "bar" -> 6)
+    }
+  }
+
+  it should "aggregate stats that still fall in the window" in testAggregate {
+    for {
+      _ <- windClock(defaultSlotDuration * 2)
+      stats <- getStats
+    } yield {
+      stats.getAll() should not be empty
+    }
+  }
+
+  it should "not aggregate beyond the window" in testAggregate {
+    for {
+      _ <- windClock(defaultSlotDuration * (defaultSlotCount + 1))
+      stats <- getStats
+    } yield {
+      stats.getAll() shouldBe empty
+    }
+  }
+
+  it should "handle 0 in configuration" in {
+    // This might happen if we base the values on something which can be 0.
+    implicit val clock = Clock.systemUTC()
+    val zeros = List(
+      TimeSlotStats[String, Int](slotDuration = 1.minutes, slotCount = 0),
+      TimeSlotStats[String, Int](slotDuration = 0.minutes, slotCount = 1)
+    )
+    Inspectors.forAll(zeros) {
+      _ shouldBe empty
+    }
+  }
+
+  it should "aggregate Int" in {
+    testRandomAggregation[String, Int](_ + _)
+  }
+
+  it should "aggregate Boolean" in {
+    implicit val boolMonoid: Monoid[Boolean] =
+      Monoid.instance[Boolean](false, _ || _)
+    testRandomAggregation[Int, Boolean](_ || _)
+  }
+
+  it should "aggregate Set" in {
+    testRandomAggregation[Int, Set[Int]](_ union _)
+  }
+
+  it should "aggregate Vector" in {
+    testRandomAggregation[Int, Vector[Int]](_ ++ _)
+  }
+
+  def testRandomAggregation[K: Arbitrary, V: Arbitrary: Monoid](f: (V, V) => V): Unit = {
+    forAll(genTimeSlotStats[K, V]) { case (stats, clock, window) =>
+      val timestamp = clock.millis()
+      val (start, end) = stats.slotRange(timestamp, window)
+
+      val windowBuffer = stats.buffer.values
+        .filter { entry =>
+          start <= entry.slotId && entry.slotId <= end
+        }
+        .toVector
+        .sortBy(_.slotId)
+
+      val all = stats.getAll(Some(window))
+
+      if (windowBuffer.exists(_.slotStats.nonEmpty)) {
+        all should not be empty
+      } else {
+        all shouldBe empty
+      }
+
+      Inspectors.forAll(all.keySet) { key =>
+        val keyStats = windowBuffer.flatMap {
+          _.slotStats.get(key)
+        }
+        val expected = keyStats.reduce(f)
+
+        all(key) shouldBe expected
+        stats.get(key, Some(window)) shouldBe expected
+      }
+    }
+  }
+}
+
+object TimeSlotStatsSpec {
+
+  class MockClock extends Clock {
+    private var currentTimeMillis = System.currentTimeMillis
+    def windByMillis(by: Long): Unit =
+      currentTimeMillis = currentTimeMillis + by
+
+    override def instant(): Instant = Instant.ofEpochMilli(currentTimeMillis)
+    override def getZone(): ZoneId = ZoneId.of("UTC")
+    override def withZone(x: ZoneId): Clock = ???
```

Comment:
pending "???"

Reply:
It will not be called, the `MockClock` doesn't need it.

Reply:
Yes, I know, it is the same case as the function “getZone()”.
Maybe in this case we can add only:
withZone(x: ZoneId): Clock = Clock.systemDefaultZone()

Reply:
I didn't check if anything needs `getZone()` internally, so I thought it's easy enough to provide some default value.

If I implemented the method the way you suggest it would not conform to the definition of the method, which is "return a copy of this clock with a different time zone", i.e. if I used `windByMillis` to set the clock to something then used `withZone` I wouldn't see the same time. I think that's more dangerous than throwing an exception that flags up clearly that this method was not expected to be called and it's not implemented, at which point you can go an implement it correctly.

In general I don't think a test-local mock implementation needs to implement methods that the test doesn't rely on, because it's just a distraction. With `???` it's clear that method can be ignored. Even mockito throws if you call methods you never set up (I tried using it initially in this test but it didn't seem to work for some reason). 

Just checked that `getZone` isn't required either, so how about turning that into `???` as well?

Reply:
Okay because it's so simple I implemented a working `withZone` to be consistent with `getZone`.

Reply:
Good! 
We have pending use of a tool from inspection of code quality (as sonarqube).
I don't sure if we can have any problems with syntax like “???”.
Anyway, good job.

Reply:
Cheers, thanks for the review!